### ISSUE TITLE
Better iCloud sync integration

### ISSUE THREAD

**Issue by ACTCD** [quoid/userscripts#424]
Currently, we tell users in the [README](https://github.com/quoid/userscripts#usage):
> Tip: for optimal cross platform experience it's a good idea to use an iCloud folder for syncing between macOS and iOS

But the current iCloud sync has some doubts for users:
- We don't really support iCloud in the app, so it won't appear in the list of Apps using iCloud.
- Not out of the box, users need to manually set and select the same path in iCloud Drive in different devices.
- The folder is not indicated as belonging to the App in iCloud (there is a folder icon containing the App Icon)
- Syncing process of script files is delayed, uncertain and uncontrolled
- For devices with limited space, local files may be evictioned due to iCloud optimization
  - Could be resolved by the `Keep Downloaded` feature introduced in macOS 15 / iOS 18
- May be briefly inaccessible due to synching, downloading, etc.; remove bookmark and restore default path

We may need to officially support iCloud features, and the synchronization mechanism and process will be managed by the App.

References:
https://developer.apple.com/library/archive/documentation/DataManagement/Conceptual/CloudKitQuickStart/Introduction/Introduction.html
https://developer.apple.com/documentation/cloudkit/
https://developer.apple.com/documentation/technotes/tn3162-understanding-cloudkit-throttles

**Comment by biuuu** [comment]
Will the data saved with GM.setValue also be synced to iCloud?

**Comment by keycommander** [comment]
Send me the link so the .js i can copy over into a scriptable application or web extension. []()

**Comment by keycommander** [comment]
> @keycommander Sorry, I don't understand what you are talking about. Can you explain it in more detail?



> @keycommander Sorry, I don't understand what you are talking about. Can you explain it in more detail?

Im just wondering how I use this repository

**Comment by 2xsaiko** [comment]
> For devices with limited space, local files may be evictioned due to iCloud optimization

I just hit this and was confused as to why my scripts suddenly weren't working anymore. Seems like Userscripts doesn't pick up the files when they get evicted until manually downloaded again through Finder. Is this known?

**Comment by 2xsaiko** [comment]
Sure, but usually, files will download on access, and non-downloaded files are visible in directories. You can see this by opening iCloud Drive in your terminal and e.g. running ls, cat or cp on a couple files. Everything is there, just accessing it might block a few seconds while it downloads, and these commands don't use the iCloud API either. Userscripts doesn't even pick the files up however.

**Comment by 2xsaiko** [comment]
Well, a simple [test program](https://gist.github.com/2xsaiko/968b3059176869e6e1b92f215bd350c8) lists the files and automatically downloads the contents, so something else is going on here.

How do you get started with this? I tried just opening the project and running the "Mac" configuration in Xcode (not even sure if that would work for a Safari extension) but it said "No such file or directory" for Resources/_locales, /dist and /images.

**Comment by 2xsaiko** [comment]
There we go: https://github.com/quoid/userscripts/pull/629

**Comment by RazorII** [comment]
I noticed now that read-only iCloud folders cannot be used as a script directory. Is this a bug. I’d really need the read-only support back so I can share a folder with others without them being able to modify the scripts.

**Comment by ACTCD** [comment]
> I noticed now that read-only iCloud folders cannot be used as a script directory. Is this a bug. I’d really need the read-only support back so I can share a folder with others without them being able to modify the scripts.

@RazorII Can you explain in more detail? You mean it was fine before? Previous and current App and OS versions?

How did you set it up to be read-only? What are the specific manifestations of failure to set?

If possible, it's a good idea to show your progress via screen recording.

**Comment by RazorII** [comment]
@ACTCD my friend just noticed their folder was reset to the default directory. When he tried to change it back to the read-only shared folder while using the latest 1.8.6 version and the open folder icon is greyed out. When I change the folder permissions to read & write then they were able to tap open and add it to Userscripts.

**Comment by ACTCD** [comment]
@RazorII Again, did read-only folders work before?

Given that opening external folders can only be done using the "Open and Save Panel", I assume this is always OS dependent and we haven't made any changes to this aspect.

**Comment by RazorII** [comment]
> [@RazorII](https://github.com/RazorII) Again, did read-only folders work before?
> 
> Given that opening external folders can only be done using the "Open and Save Panel", I assume this is always OS dependent and we haven't made any changes to this aspect.

Yes it has worked ever since I've been using Userscripts. Now I checked on their iPad and it worked there even when set to read only. Then I changed the directory on the iPhone and when I tried to change it back to the read-only folder, it didn't work again. Maybe it's an iOS bug?

**Comment by dotext** [comment]
> We may need to officially support iCloud features, and the synchronization mechanism and process will be managed by the App.

Would like to +1 for this request, as I tried to use iCloud Drive as a location for the scripts (with "Keep Downloaded" enabled) and had a few issues on both macOS and iOS, including the same issue described in #873 where the scripts were visible in the extension pop up but were not actually being applied to the page, as well as the extension pop up displaying the loading animation indefinitely at other times.

Once I deleted the app on macOS and iOS (and related folders/files on macOS), and had Userscripts use the default local location again, these issues stopped.

This was with Userscripts v5.0.0-beta.22+20260406 via TestFlight, iOS and macOS 26.5 Beta 3.

**Comment by ACTCD** [comment]
@dotext In my ongoing backend refactoring, the loading of files in the iCloud Drive directory has been addressed, so the related issues should be resolved in coming future. But we probably won't do deep integration like CloudKit.

> including the same issue described in https://github.com/quoid/userscripts/issues/873 where the scripts were visible in the extension pop up but were not actually being applied to the page, as well as the extension pop up displaying the loading animation indefinitely at other times.

Even so, if you could provide the logs when you reproduce the issues, it will still help us troubleshoot possible errors.
